### DISCUSSION TITLE
Questions about type of list elements

### DISCUSSION THREAD

**Discussion by prim** [microsoft/pyright#7765] [Q&A] [answered]
```python

class Button(object):

	def abcde(self):
		pass

l: list[Button] = [Button() for _ in range(8)]

def f(a):
	l[a].abcde()
	l[0].abcde()

```

pyright will output `b.py:11:2 - error: Type of "abcde" is unknown (reportUnknownMemberType)`.

There are two lines of .abcde() calls, one of which is correctly recognized as a Button, and the other is not.
Why not make both lines recognized as Buttons?

**Comment by rayansostenes** [comment, marked as answer]
Pyright does not know the type of the variable `a` in the function `f(a)`, so it cannot infer the type that `l[a]` will return.

For example, `a` could be a `slice`, in which case `l[a]` would return a new `list`, not a `Button` and the code would fail at runtime.
```python
f(slice(0, 2)) # AttributeError: 'list' object has no attribute 'abcde'
```

If you add a type hint to the function `f(a: int)` or `f(a: SupportsIndex)`, pyright will be able to infer the type of `l[a]` and recognize it as a Button.

```python
class Button(object):

	def abcde(self):
		pass

l: list[Button] = [Button() for _ in range(8)]

def f(a: int):
	l[a].abcde() # <- No error here
	l[0].abcde()
```

[Pyright Playground With Code](https://pyright-play.net/?pythonVersion=3.12&strict=true&locale=en&code=MYGwhgzhAEBCCuAXRB7AdgChQIwFYFNhEBKALgChyBIAE3wDNoxtg6MJ8R6zqqAHSBEohS0EAEsIiANoJk6ALrQAvNFlJUmYtHooATtAD60cWmh6waAOb4MADmILKdRvQxhRpkhSohpYBQA6ZlZbYmo-AAYgkLZiIA)

This error will only appear when the `strict` option is enabled in the pyright configuration, otherwise, pyright tends to treat `Unknown` as `Any` and does not show an error.
